File: src/Crawler/Result/Summary/Summary.php
<?php

/*
 * This file is part of the SiteOne Crawler.
 *
 * (c) Ján Regeš <jan.reges@siteone.cz>
 */

declare(strict_types=1);

namespace Crawler\Result\Summary;

use Crawler\Utils;

class Summary
{

    /**
     * @var Item[]
     */
    private array $items = [];

    /**
     * @param Item $item
     * @return void
     */
    public function addItem(Item $item): void
    {
        $this->items[] = $item;
    }

    /**
     * @return Item[]
     */
    public function getItems(): array
    {
        return $this->items;
    }

    private function sortItems(): void
    {
        usort($this->items, function (Item $a, Item $b) {
            return ItemStatus::getSortOrder($a->status) <=> ItemStatus::getSortOrder($b->status);
        });
    }

    public function getAsHtml(): string
    {
        $result = "<ul>\n";
        $this->sortItems();
        foreach ($this->items as $item) {
            $result .= '    <li>' . $item->getAsHtml() . "</li>\n";
        }
        $result .= '</ul>';
        return $result;

    }

    public function getAsConsoleText(): string
    {
        $title = "Summary";
        $titleOutput = $title . PHP_EOL . str_repeat('-', mb_strlen($title)) . PHP_EOL . PHP_EOL;;
        $result = Utils::getColorText($titleOutput, 'blue');

        $this->sortItems();
        foreach ($this->items as $item) {
            $result .= $item->getAsConsoleText() . "\n";
        }
        return $result;
    }

    public function getCountByItemStatus(ItemStatus $status): int
    {
        $count = 0;
        foreach ($this->items as $item) {
            if ($item->status === $status) {
                $count++;
            }
        }
        return $count;
    }

}

File: src/components/summary/summary.rs
// SiteOne Crawler - Summary
// (c) Jan Reges <jan.reges@siteone.cz>

use serde::{Deserialize, Serialize};

use crate::components::summary::item::Item;
use crate::components::summary::item_status::ItemStatus;
use crate::utils;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Summary {
    items: Vec<Item>,
}

impl Summary {
    pub fn new() -> Self {
        Self { items: Vec::new() }
    }

    pub fn add_item(&mut self, item: Item) {
        self.items.push(item);
    }

    pub fn get_items(&self) -> &[Item] {
        &self.items
    }

    fn sort_items(&mut self) {
        self.items.sort_by_key(|item| item.status.sort_order());
    }

    pub fn get_as_html(&mut self) -> String {
        let mut result = String::from("<ul>\n");
        self.sort_items();
        for item in &self.items {
            result.push_str(&format!("    <li>{}</li>\n", item.get_as_html()));
        }
        result.push_str("</ul>");
        result
    }

    pub fn get_as_console_text(&mut self) -> String {
        let title = "Summary";
        let title_output = format!("{}\n{}\n\n", title, "-".repeat(title.len()));
        let mut result = utils::get_color_text(&title_output, "blue", false);

        self.sort_items();
        for item in &self.items {
            result.push_str(&item.get_as_console_text());
            result.push('\n');
        }
        result
    }

    pub fn get_count_by_item_status(&self, status: ItemStatus) -> usize {
        self.items.iter().filter(|item| item.status == status).count()
    }
}
